src/AI/JakeMan.java

```diff
@@ -0,0 +1,770 @@
+      Map<UnitModel, ArrayList<XYCoord>> targetMap = new HashMap<UnitModel, ArrayList<XYCoord>>();
+
+      // Categorize all enemies by type, and all types by how well we match up vs them
+      final ArrayList<XYCoord> allEnemies = AIUtils.findEnemyUnits(myArmy, gameMap);
+      Utils.sortLocationsByTravelTime(unit, allEnemies, gameMap);
+      for( XYCoord xyc : allEnemies )
+      {
+        Unit target = gameMap.getResident(xyc);
+        UnitModel model = target.model;
+        XYCoord targetCoord = new XYCoord(target.x, target.y);
+        if (Utils.findShortestPath(unit, targetCoord, gameMap, true) != null &&
+            isWeakTo(target.model, unit.model))
+        {
+          valueMap.put(model, (double)target.getCost());
+          if (!targetMap.containsKey(model)) targetMap.put(model, new ArrayList<XYCoord>());
+          targetMap.get(model).add(targetCoord);
+        }
+      }
+
+      // Sort all individual target lists by distance
+      for (ArrayList<XYCoord> targetList : targetMap.values())
+        Utils.sortLocationsByTravelTime(unit, targetList, gameMap);
+
+      // Sort all target types by how much we want to shoot them with this unit
+      Queue<Entry<UnitModel, Double>> targetTypesInOrder = 
+          new PriorityQueue<Entry<UnitModel, Double>>(myArmy.cos[0].unitModels.size(), new UnitModelFundsComparator());
+      targetTypesInOrder.addAll(valueMap.entrySet());
+
+      while (!targetTypesInOrder.isEmpty())
+      {
+        UnitModel model = targetTypesInOrder.poll().getKey(); // peel off the juiciest
+        goals.addAll(targetMap.get(model)); // produce a list ordered by juiciness first, then distance TODO: consider a holistic "juiciness" metric that takes into account both matchup and distance?
+      }
+    }
+
+    if( goals.isEmpty() ) // Send 'em at production facilities if they haven't got anything better to do
+    {
+      for( XYCoord coord : futureCapTargets )
+      {
+        MapLocation loc = gameMap.getLocation(coord);
+        if( unit.CO.unitProductionByTerrain.containsKey(loc.getEnvironment().terrainType)
+            && myArmy.isEnemy(loc.getOwner()) )
+        {
+          goals.add(coord);
+        }
+      }
+    }
+
+    if( goals.isEmpty() ) // If there's really nothing to do, go to MY HQ
+      goals.addAll(myArmy.HQLocations);
+
+    Utils.sortLocationsByTravelTime(unit, goals, gameMap);
+    return goals;
+  }
+
+  /** Functions as working memory to prevent eviction cycles */
+  private transient Set<Unit> evictionStack;
+  private static final int    EVICTION_STACK_MAX_DEPTH = 7;
+  /**
+   * Queue the first action required to move a unit out of the way
+   * For use after unit building is complete
+   * Can recurse based on the other functions it calls.
+   */
+  private GameAction evictUnit(
+                        GameMap gameMap,
+                        Unit evicter, Unit unit,
+                        boolean avoidProduction )
+  {
```

Comment:
I ran JakeMan vs Muriel on TestRange to test the whole "will JakeMan shoot terrain with unsuitable units" hypothesis (well, first I ran JakeMan vs JakeMan, but the first one immediately capped the second one's Lab because the second one was too busy with cap chains to care, so then I put `Muriel` in slot 2. But anyway, that's a separate issue), and found that JakeMan was on the cusp of winning, but then choked himself with infantry, and slowly crumbled due to an apparent inability to clear production facilities to build more units despite banking tons of funds. Not sure what the deal was, because there were opportunities to clear production, but it seemed like he just stalled.
It may not even be an `evictUnit` problem if something was breaking before that; I haven't attempted to chase down specifically what was happening.

Reply:
> the second one was too busy with cap chains to care

I don't care about that at all. `JakeMan` isn't set up to deal with cheese.
> then choked himself with infantry, and slowly crumbled due to an apparent inability to clear production facilities to build more units despite banking tons of funds

Yeah, this is a problem common to both of my "fancy" AIs. Wally is complicated enough there's a high barrier to entry to start really messing with him, and I dunno that I'm really worried about this case for JakeMan since it's unlikely to show up in PvP maps.

Reply:
Fair enough.